### PR TITLE
Fix #740 by representing times in steps in step_current_generator.

### LINKED ISSUE #740: step_current_generator must require sufficient spacing of amplitude_times
Currently, `amplitude_times` in `step_current_generator` can be set to arbitrary doubles; NEST only checks that they are non-descending. If times are less than resolution apart, the `step_current_generator` will [not advance correctly through the amplitude array](https://github.com/nest/nest-simulator/blob/08e830bfd54943809c76b89e7ddced5f6c8ea5d7/models/step_current_generator.cpp#L245), as it can move at most one array position per time step. 

Given that the output current of the generator can only change at time steps, allowing the user to set `amplitude_times` values that do not coincide with time steps, is misleading.

Therefore, I propose that the generator in the future should only accept `amplitude_times` values that are strictly increasing multiples of the resolution and that are cast to time steps using `Time::ms_stamp()`.

The following example runs once at a high resolution, resolving all amplitude times and once at a low resolution:
```
/run_sim [ /doubletype ]  % resolution, neg power of two
{
  /res Set
  ResetKernel
  0 << /tics_per_ms 1024. /resolution res >> SetStatus
  
  /scg /step_current_generator 
    << /amplitude_times  [ 1. 1.0625 1.125 1.1875 1.25 2. 2.5 ]
       /amplitude_values [ 10. -20.  40.   -80.   160. -320. 640. ] 
    >> Create def
  /nrn /iaf_psc_alpha << /V_th 1e10 /C_m 1.0 /tau_m 1.0
                         /E_L 0.0 /V_m 0.0 /V_reset 0.0 >> Create def
  
  /vm  /voltmeter Create def
  vm << /start 1.75 /stop 4.0 /interval 0.25 >> SetStatus
  
  scg nrn Connect
  vm  nrn Connect

  5 Simulate
  vm [ /events /V_m ] get
}
def

(High resolution: ) =only 2. -5 pow run_sim ==
(Low resolution : ) =only 2. -2 pow run_sim ==
```
yielding
```
High resolution: <. 0.000000e+00 -3.137381e+00 3.294848e+01 6.105218e+01 8.293936e+01 -6.190513e+00 -7.560493e+01 8.268632e+01 2.059637e+02 .>
Low resolution : <. 0.000000e+00 2.211992e+00 -2.701283e+00 -6.527746e+00 -9.507798e+00 -1.182866e+01 -1.363616e+01 1.309476e+02 2.435496e+02 .>
```

### PR DESCRIPTION
**Pull request by heplesser**
@appukuttan-shailesh Could you be the second reviewer for this PR?

### CODE REVIEW

**Comment by appukuttan-shailesh** [comment]
Sure, I can take this up.

File: models/step_current_generator.cpp [hunk 1]
```diff
@@ -100,48 +108,126 @@ nest::step_current_generator::Buffers_::Buffers_( const Buffers_&,
...
+  ( *d )[ names::allow_offgrid_times ] = BoolDatum( allow_offgrid_amp_times_ );
+}
+
+nest::Time
+nest::step_current_generator::Parameters_::validate_time_( double t,
+  const Time& t_previous )
+{
+  if ( t == 0.0 )
+  {
+    throw BadProperty( "Amplitude cannot be changed at 0." );
```
Comment (appukuttan-shailesh) [inline models/step_current_generator.cpp]: Preferably update to "Amplitude cannot be changed at t = 0."

Comment (appukuttan-shailesh) [inline models/step_current_generator.cpp]: Preferably update to "Amplitude cannot be changed at t = 0."

Comment (heplesser) [reply models/step_current_generator.cpp]: Fixed.

File: models/step_current_generator.cpp [hunk 2]
```diff
@@ -100,48 +108,126 @@ nest::step_current_generator::Buffers_::Buffers_( const Buffers_&,
...
+    {
+      // In this case, we need to round to the end of the step
+      // in which t lies, ms_stamp does that for us.
+      t_amp = Time::ms_stamp( t );
+    }
+    else
+    {
+      std::stringstream msg;
+      msg << "step_current_generator: Time point " << t
+          << " is not representable in current resolution.";
```
[the remark this answers is not in the document]
Comment (heplesser) [reply models/step_current_generator.cpp]: I decided to combine this with the t==0 check: if t <= 0, the user now gets the following message: "Amplitude can only be changed at strictly positive times (t > 0)."

**Comment by heplesser** [comment]
@jougs One use case would be that the user uses random numbers for the times at which amplitudes change. These would not most likely not fall onto the grid. As a convenience for the user, NEST can round these times, but it should only do this if explicitly requested to avoid subtle errors otherwise. 

Should I put this into the inline documentation, or did you just want an explanation here in the discussion?

**Comment by jougs** [comment]
@heplesser: thanks for the explanation. I think it won't hurt if you added something along those lines to the documentation. Thanks for the nice work!

**Comment by heplesser** [comment]
@jougs Done, hope it is ready to merge now!

**Pull request merged**